nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -0,0 +1,504 @@
+package deploymentwatcher
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// evalBatchPeriod is the batching length before creating an evaluation to
+	// trigger the scheduler when allocations are marked as healthy.
+	evalBatchPeriod = 1 * time.Second
+)
+
+// deploymentTriggers are the set of functions required to trigger changes on
+// behalf of a deployment
+type deploymentTriggers interface {
+	// createEvaluation is used to create an evaluation.
+	createEvaluation(eval *structs.Evaluation) (uint64, error)
+
+	// upsertJob is used to roll back a job when autoreverting for a deployment
+	upsertJob(job *structs.Job) (uint64, error)
+
+	// upsertDeploymentStatusUpdate is used to upsert a deployment status update
+	// and an optional evaluation and job to upsert
+	upsertDeploymentStatusUpdate(u *structs.DeploymentStatusUpdate, eval *structs.Evaluation, job *structs.Job) (uint64, error)
+
+	// upsertDeploymentPromotion is used to promote canaries in a deployment
+	upsertDeploymentPromotion(req *structs.ApplyDeploymentPromoteRequest) (uint64, error)
+
+	// upsertDeploymentAllocHealth is used to set the health of allocations in a
+	// deployment
+	upsertDeploymentAllocHealth(req *structs.ApplyDeploymentAllocHealthRequest) (uint64, error)
+}
+
+// deploymentWatcher is used to watch a single deployment and trigger the
+// scheduler when allocation health transistions.
+type deploymentWatcher struct {
+	// queryLimiter is used to limit the rate of blocking queries
+	queryLimiter *rate.Limiter
+
+	// deploymentTriggers holds the methods required to trigger changes on behalf of the
+	// deployment
+	deploymentTriggers
+
+	// DeploymentStateWatchers holds the methods required to watch objects for
+	// changes on behalf of the deployment
+	DeploymentStateWatchers
+
+	// d is the deployment being watched
+	d *structs.Deployment
```

Comment:
>  the further from its declaration that a name is used, the more descriptive the name must be

I would argue these declarations are far enough from their uses to warrant at least 2 letters ;)